pkg/kafka_ingester/kafka_ingester.go

```diff
@@ -0,0 +1,255 @@
+			CommitInterval: config.CommitInterval,
+			RetentionTime:  config.RetentionTime,
+			StartOffset:    startOffset,
+			Logger:         kafkaLogger,
+			ErrorLogger:    kafkaErrorLogger,
+		})
+
+	kafkaConnectionInfo.With(prometheus.Labels{
+		"brokers":  strings.Join(config.Brokers, ","),
+		"group_id": config.GroupID,
+		"topic":    config.Topic,
+	}).Set(1)
+
+	return &KafkaIngester{
+		outputChannel:   make(chan *event.Raw),
+		shutdownChannel: make(chan struct{}),
+		done:            false,
+		logger:          logger,
+		kafkaReader:     reader,
+	}, nil
+}
+
+func (k *KafkaIngester) RegisterEventProcessingDurationObserver(observer pipeline.EventProcessingDurationObserver) {
+	k.observer = observer
+}
+
+func (k *KafkaIngester) observeDuration(start time.Time) {
+	if k.observer != nil {
+		k.observer.Observe(time.Since(start).Seconds())
+	}
+}
+
+func (k *KafkaIngester) RegisterMetrics(_ prometheus.Registerer, wrappedRegistry prometheus.Registerer) error {
+	toRegister := []prometheus.Collector{kafkaConnectionInfo, messagesReadTotal, malformedMessagesTotal}
+	for _, collector := range toRegister {
+		if err := wrappedRegistry.Register(collector); err != nil {
+			return fmt.Errorf("error registering metric %s: %w", collector, err)
+		}
+	}
+	return nil
+}
+
+func (k *KafkaIngester) Done() bool {
+	return k.done
+}
+
+func (k *KafkaIngester) OutputChannel() chan *event.Raw {
+	return k.outputChannel
+}
+
+// Run starts to tail the associated file, feeding events to output channel.
+func (k *KafkaIngester) Run() {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Goroutine handling shutdown signal
+	go func() {
+		defer func() {
+			close(k.outputChannel)
+			k.done = true
+		}()
+		for {
+			select {
+			case <-k.shutdownChannel:
+				cancel()
+				k.kafkaReader.Close()
+				return
+			}
+		}
+	}()
+
+	// Main goroutine for reading messages from Kafka
+	go func() {
+		for {
+			m, err := k.kafkaReader.ReadMessage(ctx)
+			start := time.Now()
+			if err != nil {
+				if err == io.EOF || err == context.Canceled {
+					return
+				}
+				k.logger.Errorf("error while reading message from Kafka: %w", err)
+			}
+			k.logger.Debug(m)
+			messagesReadTotal.Inc()
+			e, err := processMessage(m)
+			if err != nil {
+				k.logger.Errorf("Error while parsing the message: %w", err)
+				malformedMessagesTotal.Inc()
+			} else {
+				k.outputChannel <- e
+			}
+			k.observeDuration(start)
+		}
+	}()
+}
+
+func getSchemaVersionFromHeaders(headers []kafka.Header) (string, bool) {
+	var (
+		found     bool
+		schemaVer string
+	)
+	for _, header := range headers {
+		if header.Key == schemaVersionMessageHeader {
+			schemaVer = string(header.Value)
+			found = true
+			break
+		}
+	}
+	return schemaVer, found
+}
+
+func processMessage(m kafka.Message) (*event.Raw, error) {
+	schemaVer, ok := getSchemaVersionFromHeaders(m.Headers)
+	if !ok {
+		schemaVer = defaultSchemaVer
+	}
+	switch schemaVer {
+	case "v1":
```

Comment:
If we are already using the constant for default version we probably should not duplicate it here? 
I'd prefer to use a constant for v1 and make the default equal to it

Reply:
d914de9